rescan_test.go

```diff
@@ -0,0 +1,680 @@
+		pprof.Lookup("goroutine").WriteTo(os.Stdout, 1)
+		ctx.t.Fatal("rescan did not exit cleanly")
+	}
+}
+
+// recvBlockDisconnected is a helper method used to expect a block connected
+// notification being delivered.
+func (ctx *rescanTestContext) recvBlockConnected(block waddrmgr.BlockStamp) {
+	ctx.t.Helper()
+
+	select {
+	case recvBlock := <-ctx.blocksConnected:
+		if !reflect.DeepEqual(recvBlock, block) {
+			ctx.t.Fatalf("expected block connected notification "+
+				"for height %v, got %v", spew.Sdump(block),
+				spew.Sdump(recvBlock))
+		}
+	case <-time.After(time.Second):
+		ctx.t.Fatalf("expected to receive block connected "+
+			"notification for %v", spew.Sdump(block))
+	}
+}
+
+// recvBlockDisconnected is a helper method used to expect a block disconnected
+// notification being delivered.
+func (ctx *rescanTestContext) recvBlockDisconnected(height int32) {
+	ctx.t.Helper()
+
+	select {
+	case recvBlock := <-ctx.blocksDisconnected:
+		if !reflect.DeepEqual(recvBlock, block) {
+			ctx.t.Fatalf("expected block disconnected notification "+
+				"for height %v, got %v", spew.Sdump(block),
+				spew.Sdump(recvBlock))
+		}
+	case <-time.After(time.Second):
+		ctx.t.Fatalf("expected to receive block disconnected "+
+			"notification for %v", spew.Sdump(block))
+	}
+}
+
+// waitUntilSynced blocks until the underlying rescan is ready to receive
+// notifications at tip.
+func (ctx *rescanTestContext) waitUntilSynced() {
+	ctx.t.Helper()
+
+	signal := make(chan struct{})
+	ctx.chain.overrideSubscribe(func(bestHeight uint32) (*blockntfns.Subscription, error) {
+		if bestHeight == uint32(ctx.chain.bestBlock.Height) {
+			close(signal)
+		}
+		return &blockntfns.Subscription{
+			Notifications: ctx.chain.ntfnChan,
+			Cancel:        func() {},
+		}, nil
+	})
+
+	select {
+	case <-signal:
+	case <-time.After(time.Second):
+		ctx.t.Fatal("expected to receive synced signal")
+	}
+
+	ctx.chain.overrideSubscribe(nil)
+}
+
+// TestRescanRetryBlocks tests the interaction of the Rescan with regards to
+// retry blocks under different scenarios.
+func TestRescanRetryBlocks(t *testing.T) {
+	t.Parallel()
+
+	// Define some helper closures to ease the ability of marking a block as
+	// a retry block. This is done by replacing the GetCFilter
+	// implementation with one that returns a nil filter, causing the Rescan
+	// to retry fetching it.
+	filterQueried := make(chan chainhash.Hash, 1)
+	filterNotFound := func(hash chainhash.Hash,
+		_ wire.FilterType, _ ...QueryOption) (*gcs.Filter, error) {
+
+		filterQueried <- hash
+		return nil, nil
+	}
+
+	assertFilterQueried := func(t *testing.T, hash chainhash.Hash) {
+		t.Helper()
+
+		var hashQueried chainhash.Hash
+		select {
+		case hashQueried = <-filterQueried:
+			if !hashQueried.IsEqual(&hash) {
+				t.Fatalf("expected to query filter %v, got %v",
+					hash, hashQueried)
+			}
+		case <-time.After(time.Second):
+			t.Fatal("expected rescan to query for filter")
+		}
+	}
```

Comment:
Shouldn't this be part of the `mockChainSource`? Make a bool field that controls `filterNotFound` behaviour. Then also all filter queries can be pushed into a channel and asserted on, not just the not found ones.

Reply:
Done.